src/libraries/Common/src/System/Text/Base64UrlEncoding.cs

```diff
@@ -0,0 +1,67 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+
+namespace System.Text
+{
+    internal static class Base64UrlEncoding
+    {
+        internal static ArraySegment<char> RentEncode(ReadOnlySpan<byte> input)
+        {
+            // Every 3 bytes turns into 4 chars for the Base64 operation
+            int base64Len = ((input.Length + 2) / 3) * 4;
+            char[] base64 = ArrayPool<char>.Shared.Rent(base64Len);
+
+            if (!Convert.TryToBase64Chars(input, base64, out int charsWritten))
+            {
+                Debug.Fail($"Convert.TryToBase64 failed with {input.Length} bytes to a {base64.Length} buffer");
+                throw new UnreachableException();
+            }
+
+            Debug.Assert(charsWritten == base64Len);
+
+            // In the degenerate case every char will turn into 3 chars.
+            int urlEncodedLen = charsWritten * 3;
+            char[] urlEncoded = ArrayPool<char>.Shared.Rent(urlEncodedLen);
+            int writeIdx = 0;
+
+            for (int readIdx = 0; readIdx < charsWritten; readIdx++)
+            {
+                char cur = base64[readIdx];
+
+                if (char.IsAsciiLetterOrDigit(cur))
+                {
+                    urlEncoded[writeIdx++] = cur;
+                }
+                else if (cur == '+')
+                {
+                    urlEncoded[writeIdx++] = '%';
+                    urlEncoded[writeIdx++] = '2';
+                    urlEncoded[writeIdx++] = 'B';
+                }
+                else if (cur == '/')
+                {
+                    urlEncoded[writeIdx++] = '%';
+                    urlEncoded[writeIdx++] = '2';
+                    urlEncoded[writeIdx++] = 'F';
+                }
+                else if (cur == '=')
+                {
+                    urlEncoded[writeIdx++] = '%';
+                    urlEncoded[writeIdx++] = '3';
+                    urlEncoded[writeIdx++] = 'D';
+                }
+                else
+                {
+                    Debug.Fail($"'{cur}' is not a valid Base64 character");
+                    throw new UnreachableException();
+                }
+            }
```

Comment:
How much data is this typically going to process?  Presumably we'd expect +, /, and = to be relatively rare, in which case we could do something along the lines of:
```C#
ReadOnlySpan<char> source = base64.AsSpan(0, charsWritten);
Span<char> dest = urlEncoded;
int written = 0;
while (!source.IsEmpty)
{
    int pos = source.IndexOfAny('+', '/', '=');
    if (pos < 0)
    {
        written += source.Length;
        source.CopyTo(dest);
        break;
    }

    source.Slice(0, pos).CopyTo(dest);
    source = source.Slice(pos);
    dest = dest.Slice(pos);
    written += pos;

    dest[0] = '%';
    switch (source[pos])
    {
        case '+': dest[1] = '2'; dest[2] = 'B'; break;
        case '/': dest[1] = '2'; dest[2] = 'F'; break;
        case '=': dest[1] = '3'; dest[2] = 'D'; break;
    }

    source = source.Slice(1);
    dest = dest.Slice(3);
}
```

Reply:
Also, why does this base64url encoding logic differ from https://github.com/dotnet/aspnetcore/blob/a849c806c01df63201e9d32289be436d1a49f315/src/Shared/WebEncoders/WebEncoders.cs#L303-L319?  That version, which doesn't expand on the replacements, means it can be done in-place and not require a second rental.  It'd also be nice to avoid renting at all if we can use stack space due to the input being short enough.

Reply:
Ah, I might be using a bad name here.  This is URLEncode(Base64(data)), which is what's required for OCSP.  That's `base64url`, a different scheme, to avoid the re-encode.